feat: add annotation to ignore local storage volume during scale down

vadasambar opened the pull request:
- this is so that scale down is not blocked on local storage volume
- for pods where it is okay to ignore local storage volume
Signed-off-by: vadasambar <surajrbanakar@gmail.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Some pods like istio sidecars have local storage which block node scale down. We would like to have the ability to ignore local storage volumes on certain pods provided the right annotation of the form `cluster-autoscaler.kubernetes.io/ignore-local-storage-volume: <volume-name>` is present on the pod. 
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #3947 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added: cluster-autoscaler now supports a new pod annotation `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes: "volume-1,volume-2"` to evict pods with local storage volumes (where `volume-1` and `volume-2` are safe to evict local storage volumes) during node scale down.  
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
TBD
```

vadasambar commented:
Test suite is failing because the tests haven't been updated for the new annotation yet.

vadasambar commented on cluster-autoscaler/utils/drain/drain.go:
```diff
@@ -38,6 +38,8 @@ const (
 	// kube-system namespace or having a local storage.
 	PodSafeToEvictKey = "cluster-autoscaler.kubernetes.io/safe-to-evict"
+	// IgnoreLocalStorageVolumeKey - annotation that ignores (doesn't block on) a local storage volume during scale down
+	IgnoreLocalStorageVolumeKey = "cluster-autoscaler.kubernetes.io/ignore-local-storage-volume"
```
Better naming suggestions are welcome.

towca replied:
"safe-to-evict" is a widely recognized concept at this point, maybe we could reuse it, e.g. `cluster-autoscaler.kubernetes.io/local-volume-safe-to-evict`? I know we're not technically evicting the volume, but I think it conveys the intent better. Or maybe include "scale down" and "blocking" instead, like `cluster-autoscaler.kubernetes.io/scale-down-non-blocking-local-volume`? Just "ignore" doesn't tell much about what purpose it's being ignored for.

On a slightly related note, I don't think we should restrict ourselves to just one volume here, what if there's more that should be ignored? WDYT about making the value of the annotation a comma-separated list of volume names instead?

vadasambar commented on cluster-autoscaler/simulator/drain_test.go:
```diff
@@ -115,6 +115,7 @@ func TestGetPodsToMove(t *testing.T) {
 		Spec: apiv1.PodSpec{
 			Volumes: []apiv1.Volume{
 				{
+					Name: "empty-vol",
```
Added volume name because it is a required field in Kubernetes. I realized the tests in this files were failing because I wasn't handling the case for empty volume name [here](https://github.com/kubernetes/autoscaler/pull/5594/commits/b0aca12b2685dfb9ffbbc98dda526f2099d9dec3#diff-a22fbdb27ac8bbd8f9c9597042678aa5f6541d207160adc775d56e13f3fe99beR259). This has been fixed in this PR [here](https://github.com/kubernetes/autoscaler/pull/5594/files#diff-a22fbdb27ac8bbd8f9c9597042678aa5f6541d207160adc775d56e13f3fe99beR271-R274).

I thought about removing the volume name above since I am handling the case for empty volume name but I kept it since this makes the manifest closer to actual manifest on Kubernetes. 

```
suraj@suraj:~/Downloads$ kubectl explain pod.spec.volumes | grep name
     Volume represents a named volume in a pod that may be accessed by any
   name	<string> -required-
     name of the volume. Must be a DNS_LABEL and unique within the pod. More
     https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
     PersistentVolumeClaim in the same namespace. More info:
```

towca replied:
Ack, that makes sense.

vadasambar commented on cluster-autoscaler/simulator/drain_test.go:
```diff
@@ -136,6 +137,7 @@ func TestGetPodsToMove(t *testing.T) {
 		Spec: apiv1.PodSpec{
 			Volumes: []apiv1.Volume{
 				{
+					Name: "my-repo",
```
Same comment as https://github.com/kubernetes/autoscaler/pull/5594/files#r1138360614

vadasambar commented:
Can I ask for a review on this before I go and test it on a live cluster :pray: 

Asking this because I think it's a small change and I think the code should work. If possible, I want to address the comments first, get an OK and go test things on a live cluster. This would save me one (or more) round(s) of manual testing :bow:

towca commented:
Hi! We already have an annotation that you can use to mark a pod as "never blocking scale down" - `cluster-autoscaler.kubernetes.io/safe-to-evict=true`. Maybe that's enough for your use-case? If not, could you share why? The "pod blocking scale-down" logic basically guesses whether it's safe to evict a given pod. If you're annotating the pod manually anyway, shouldn't you be able to tell if it's generally safe to evict (so there's no need to guess)?

towca commented on cluster-autoscaler/utils/drain/drain_test.go:
```diff
@@ -489,6 +509,16 @@ func TestDrain(t *testing.T) {
 			expectBlockingPod:   &BlockingPod{Pod: emptydirPod, Reason: LocalStorageRequested},
 			expectDaemonSetPods: []*apiv1.Pod{},
 		},
+		{
```
I'd consider also including test cases for pod with local volume(s) and:
* The annotation value being empty (=no local volumes match, pod is blocking)
* The annotation value not matching any local volumes on the pod (=pod is blocking)
* The annotation value only matching one local volume on the pod, but there is another one without the annotation (=pod is still blocking)
* [if we have multiple values available] Multiple annotation values matching all local volumes on the pod (=pod is not blocking)

vadasambar replied:
All of them make sense. I came up with a couple more after reading your comment.

vadasambar replied:
I have added the new test cases. Thank you.

towca reviewed (changes requested):
Ahh I get it, it makes a lot of sense for sidecars, thanks for the explanation. Overall the PR looks good, I just had a couple comments. Also, could you please squash the intermediate commits? I think this is all pretty self-contained and 1 commit for the whole PR should suffice.

vadasambar commented:
> "safe-to-evict" is a widely recognized concept at this point, maybe we could reuse it, e.g. `cluster-autoscaler.kubernetes.io/local-volume-safe-to-evict`? I know we're not technically evicting the volume, but I think it conveys the intent better. Or maybe include "scale down" and "blocking" instead, like `cluster-autoscaler.kubernetes.io/scale-down-non-blocking-local-volume`? Just "ignore" doesn't tell much about what purpose it's being ignored for.

I think `safe-to-evict` conveys the intent better than `ignore`. `cluster-autoscaler.kubernetes.io/scale-down-non-blocking-local-volume` is good to but when I read it I thought of 2 possible readings of the annotation:
1. `scale-down` the `non-blocking-local-volume`
2. `scale-down-non-blocking` (for node)  `local-volume`

`ignore-local-storage-volume` also has the same vagueness about it e.g., does it mean ignore storage volume during scale down or ignore local storage volume for something else?

I think `safe-to-evict-local-volume` might be slightly easier to remember than `local-volume-safe-to-evict`. WDYT?

vadasambar commented:
> On a slightly related note, I don't think we should restrict ourselves to just one volume here, what if there's more that should be ignored? WDYT about making the value of the annotation a comma-separated list of volume names instead?

I like the idea. :+1:

vadasambar commented:
> Also, could you please squash the intermediate commits?

Wanted to squash but after seeing your comments I think I need to add more commits. Do you mind if I do squashing at the end before we merge the PR?

vadasambar commented on cluster-autoscaler/utils/drain/drain.go:
```diff
@@ -38,6 +39,8 @@ const (
 	// kube-system namespace or having a local storage.
 	PodSafeToEvictKey = "cluster-autoscaler.kubernetes.io/safe-to-evict"
+	// SafeToEvictLocalVolumeKey - annotation that ignores (doesn't block on) a local storage volume during node scale down
+	SafeToEvictLocalVolumeKey = "cluster-autoscaler.kubernetes.io/safe-to-evict-local-volume"
```
I have changed the annotation key here to `safe-to-evict-local-volume` instead of the suggested `local-volume-safe-to-evict` because I think it's easier to read and remember. This is of course my thinking. Happy to change it based on review comments.

vadasambar replied:
Annotation key name is being discussed in this comment: https://github.com/kubernetes/autoscaler/pull/5594#issuecomment-1486257828

vadasambar replied:
Makes sense :+1:

vadasambar replied:
Updated to `safe-to-evict-local-volumes`

vadasambar commented on cluster-autoscaler/utils/drain/drain.go:
```diff
@@ -250,10 +253,19 @@ func isPodTerminal(pod *apiv1.Pod) bool {
+	if annotationVal != "" {
+		vols := strings.Split(annotationVal, ",")
+		for _, v := range vols {
+			isNonBlocking[v] = true
```
I am creating a map here which seems a little wasteful. Any better suggestions are welcome :pray:

vadasambar replied:
:memo: thank you for the explanation. 

Say we have a lot of pods which need to be drained from a node marked for scale down. Would it be fair to say, since we drain only one node at a time, the checking we do for every pod is still not as performance heavy as API calls and binpacking complexity.

towca replied:
Yes exactly, it should be negligible in practice. Also, k8s imposes limits on the maximum number of pods on a node (default is 110 I think? At least in GKE it's somewhere in that ballpark), and the actual number is in practice much lower than the limit, due to many reasons.

Also FYI, we have parallel drain in the works, it'll probably be turned on by default in the next CA release.

vadasambar commented on cluster-autoscaler/utils/drain/drain_test.go:
```diff
@@ -210,6 +210,150 @@ func TestDrain(t *testing.T) {
+		},
+	}
+
+	emptydirPodWithSafeToEvictLocalVolAnnotationMultiValAllMatching := &apiv1.Pod{
```
Test names here e.g., `emptydirPodWithSafeToEvictLocalVolAnnotationMultiValAllMatching` feel a little too long. Any better suggestions are welcome :pray:

towca replied:
`emptyDirSafeToEvictVolumeMultiValAllMatching` etc.? Seems to have more or less the same information.

vadasambar replied:
> `emptyDirSafeToEvictVolumeMultiValAllMatching` etc.? Seems to have more or less the same information.

I like this better. Thank you!

vadasambar replied:
Updated the names.

vadasambar commented:
@towca I think I have addressed all the review comments from my side. Waiting for your review.

towca reviewed (changes requested):
Thanks for the changes, looks good overall. Just one comment about pluralizing the annotation name + one nit.

Also, could you document this annotation somewhere in here: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node ?

vadasambar commented on cluster-autoscaler/utils/drain/drain.go:
```diff
@@ -250,10 +253,19 @@ func isPodTerminal(pod *apiv1.Pod) bool {
+// HasBlockingLocalStorage returns true if pod has any local storage
+// without pod annotation `<SafeToEvictLocalVolumeKey>: <volume-name-1>,<volume-name-2>...`
+func HasBlockingLocalStorage(pod *apiv1.Pod) bool {
+	annotationVal := pod.GetAnnotations()[SafeToEvictLocalVolumeKey]
```
I like the idea. Thank you for the suggestion!

vadasambar replied:
I have abstracted it out in a separate function.

vadasambar commented:
> Thanks for the changes, looks good overall. Just one comment about pluralizing the annotation name + one nit.
> 
> Also, could you document this annotation somewhere in here: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node ?

I missed adding info about this in the FAQ. Thank you for bringing this up. :pray:

towca commented:
The code looks great, thanks for the contribution!

Before merging, could you:
* Add the FAQ changes to this PR
* Update the "Does this PR introduce a user-facing change?" section of the PR (we use it to compile change logs while releasing CA) to mention the correct annotation and expand on what the volumes are ignored for (basically that they're not blocking eviction).
* Squash the commits to just 1

vadasambar commented:
> The code looks great, thanks for the contribution!
> 
> Before merging, could you:
> 
>   * Add the FAQ changes to this PR
> 
>   * Update the "Does this PR introduce a user-facing change?" section of the PR (we use it to compile change logs while releasing CA) to mention the correct annotation and expand on what the volumes are ignored for (basically that they're not blocking eviction).
> 
>   * Squash the commits to just 1

@towca thank you for reviewing my PR. I have added the FAQ, updated the user facing change section of the PR and squashed everything into 1 commit. Since the FAQ and user facing change section is user-facing if you have any suggestions to make it better please let me know :pray:. Happy to change them and re-squash the commits again (and thank you for the great suggestions).

towca commented:
The PR description looks great, the wording in the FAQ does as well, but the formatting is getting confusing there. I added a comment with some ideas on how to format it better, after that we can merge.

towca commented on cluster-autoscaler/FAQ.md:
```diff
@@ -88,7 +88,15 @@ Cluster Autoscaler decreases the size of the cluster when some nodes are consist
 * Pods that are not backed by a controller object (so not created by deployment, replica set, job, stateful set etc). *
-* Pods with local storage. *
+* Pods with local storage. *  
+Unless the pod has the following annotation set:
```
Including the new info in this place unfortunately breaks the list which looks pretty confusing.

Sadly this point already has an asterisk that other points also use, so we can't piggyback on it.

I'd either:
* Add a second asterisk and include the information below the whole list, like we do for `cluster-autoscaler.kubernetes.io/safe-to-evict`.
* Shorten the information - use in-line code tags instead of the blocks, remove the clarification about using only one volume - if you say "a comma-separated list" I think it's pretty widely understood that a single item is also valid. Add the shortened info to the list in a sub-bullet - like we do for kube-system pods above.

The second one seems slightly better as the asterisks might get confusing.

vadasambar replied:
Thank you for the suggestion. The spacing for the last part does look different from the bullet. 

> Add a second asterisk and include the information below the whole list, like we do for cluster-autoscaler.kubernetes.io/safe-to-evict.

I kind of like this idea because we'd be able to write:
```
"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "volume-1,volume-2"
```
as opposed to `
"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "volume-1,volume-2"`. First option is easier to catch the eye because it is a block. 

We can get over the asterisk problem with something like this:

> 
> * Pods with local storage. * # 
> ...
>
> <sup>*</sup>Unless the pod has the following annotation (supported in CA 1.0.3 or later):
> ```
> "cluster-autoscaler.kubernetes.io/safe-to-evict": "true"
> ```
> __Or__
> 
> <sup>#</sup>Unless the pod has an annotation like:
> ```
> "cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "volume-1,volume-2"
> ```
> where `volume-1` and `volume-2` are safe to evict local storage volumes. You can also specify only one volume without any comma like this:
> ```
> "cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "volume-1"
> ```
> 
> __Or__ you have...

> Shorten the information - use in-line code tags instead of the blocks, remove the clarification about using only one volume - if you say "a comma-separated list" I think it's pretty widely understood that a single item is also valid.

I think this would be less confusing at the cost of being less visible.

vadasambar replied:
> if you say "a comma-separated list" I think it's pretty widely understood that a single item is also valid.

True. I just wanted to be explicit here :sweat_smile:

vadasambar replied:
P.S. I tried something and I think it should solve the formatting problem without us having to do 1 or 2 above. Just adding a few spaces to the line aligns it nicely with the bullet. Here's what I mean:
![image](https://user-images.githubusercontent.com/34534103/230541607-c37573a1-ccae-4082-a4f1-eafecf947a6f.png)

vadasambar replied:
I will try this and push. Let me know what you think.

vadasambar replied:
Added. https://github.com/kubernetes/autoscaler/pull/5594/files#r1160432848

vadasambar commented on cluster-autoscaler/FAQ.md:
```diff
@@ -88,7 +88,12 @@ Cluster Autoscaler decreases the size of the cluster when some nodes are consist
 * Pods that are not backed by a controller object (so not created by deployment, replica set, job, stateful set etc). *
-* Pods with local storage. *
+* Pods with local storage. *  
+    - unless the pod has the following annotation set:
```
I took the liberty of adding a sub-bullet here. Happy to change if you have any feedback.

vadasambar replied:
The preview is a little messed up. Here's what it actually looks like:
![image](https://user-images.githubusercontent.com/34534103/230542802-b4d43105-e268-4cb2-9a4d-9d765a23ae04.png)
https://github.com/kubernetes/autoscaler/blob/df1b7e73e73ac619c83623588421096b3f19ddb8/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node

towca replied:
The formatting looks good.

nit: I'd change the wording from "where..." to "and all of the pod's local volumes are listed in the annotation value", it's a bit more accurate.

vadasambar replied:
I think it makes things clearer too. Thank you for the suggestion. Will update the PR.

vadasambar commented on cluster-autoscaler/utils/drain/drain_test.go:
```diff
@@ -489,6 +661,76 @@ func TestDrain(t *testing.T) {
+			expectDaemonSetPods: []*apiv1.Pod{},
+		},
+		{
+			description:         "pod with EmptyDir and SafeToEvictLocalVolumesKey annotation with non-matching values",
```
Added an extra test here to cover the following case:
1. Pod has volumes `scratch-1,scratch-2,scratch-3` but user specifies `scratch-1,scratch-2,scratch-5` in the annotation

vadasambar commented:
### Testing
#### Without the annotation
![image](https://user-images.githubusercontent.com/34534103/231360865-1e03560a-b1ff-4106-8852-2861ed01fd19.png)

<details>
<summary>
Here's what the pod looks like
</summary>

```
suraj@suraj:~/bin$ k get po nginx-777b5694f6-hfvx6 -oyaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubectl.kubernetes.io/default-container: nginx
    kubectl.kubernetes.io/default-logs-container: nginx
    prometheus.io/path: /stats/prometheus
    prometheus.io/port: "15020"
    prometheus.io/scrape: "true"
    sidecar.istio.io/status: '{"initContainers":["istio-init"],"containers":["istio-proxy"],"volumes":["workload-socket","credential-socket","workload-certs","istio-envoy","istio-data","istio-podinfo","istio-token","istiod-ca-cert"],"imagePullSecrets":null,"revision":"default"}'
  creationTimestamp: "2023-04-12T05:27:23Z"
  generateName: nginx-777b5694f6-
  labels:
    app: nginx
    pod-template-hash: 777b5694f6
    security.istio.io/tlsMode: istio
    service.istio.io/canonical-name: nginx
    service.istio.io/canonical-revision: latest
  name: nginx-777b5694f6-hfvx6
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: nginx-777b5694f6
    uid: 83078dec-aedd-433e-bde3-4f6ed87540f0
  resourceVersion: "35993"
  uid: ce4c95f7-7a5d-4e8f-820e-e40e901cb54e
spec:
  containers:
  - image: nginx
    imagePullPolicy: Always
    name: nginx
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-wknsc
      readOnly: true
  - args:
    - proxy
    - sidecar
    - --domain
    - $(POD_NAMESPACE).svc.cluster.local
    - --proxyLogLevel=warning
    - --proxyComponentLogLevel=misc:error
    - --log_output_level=default:info
    - --concurrency
    - "2"
    env:
    - name: JWT_POLICY
      value: third-party-jwt
    - name: PILOT_CERT_PROVIDER
      value: istiod
    - name: CA_ADDR
      value: istiod.istio-system.svc:15012
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: INSTANCE_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    - name: SERVICE_ACCOUNT
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.serviceAccountName
    - name: HOST_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.hostIP
    - name: PROXY_CONFIG
      value: |
        {}
    - name: ISTIO_META_POD_PORTS
      value: |-
        [
        ]
    - name: ISTIO_META_APP_CONTAINERS
      value: nginx
    - name: ISTIO_META_CLUSTER_ID
      value: Kubernetes
    - name: ISTIO_META_NODE_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
    - name: ISTIO_META_INTERCEPTION_MODE
      value: REDIRECT
    - name: ISTIO_META_WORKLOAD_NAME
      value: nginx
    - name: ISTIO_META_OWNER
      value: kubernetes://apis/apps/v1/namespaces/default/deployments/nginx
    - name: ISTIO_META_MESH_ID
      value: cluster.local
    - name: TRUST_DOMAIN
      value: cluster.local
    image: docker.io/istio/proxyv2:1.17.1
    imagePullPolicy: IfNotPresent
    name: istio-proxy
    ports:
    - containerPort: 15090
      name: http-envoy-prom
      protocol: TCP
    readinessProbe:
      failureThreshold: 30
      httpGet:
        path: /healthz/ready
        port: 15021
        scheme: HTTP
      initialDelaySeconds: 1
      periodSeconds: 2
      successThreshold: 1
      timeoutSeconds: 3
    resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 10m
        memory: 40Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsGroup: 1337
      runAsNonRoot: true
      runAsUser: 1337
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/workload-spiffe-uds
      name: workload-socket
    - mountPath: /var/run/secrets/credential-uds
      name: credential-socket
    - mountPath: /var/run/secrets/workload-spiffe-credentials
      name: workload-certs
    - mountPath: /var/run/secrets/istio
      name: istiod-ca-cert
    - mountPath: /var/lib/istio/data
      name: istio-data
    - mountPath: /etc/istio/proxy
      name: istio-envoy
    - mountPath: /var/run/secrets/tokens
      name: istio-token
    - mountPath: /etc/istio/pod
      name: istio-podinfo
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-wknsc
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  initContainers:
  - args:
    - istio-iptables
    - -p
    - "15001"
    - -z
    - "15006"
    - -u
    - "1337"
    - -m
    - REDIRECT
    - -i
    - '*'
    - -x
    - ""
    - -b
    - '*'
    - -d
    - 15090,15021,15020
    - --log_output_level=default:info
    image: docker.io/istio/proxyv2:1.17.1
    imagePullPolicy: IfNotPresent
    name: istio-init
    resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 10m
        memory: 40Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        add:
        - NET_ADMIN
        - NET_RAW
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: false
      runAsGroup: 0
      runAsNonRoot: false
      runAsUser: 0
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-wknsc
      readOnly: true
  nodeName: gke-cluster-1-default-pool-87aace2f-vwr9
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir: {}
    name: workload-socket
  - emptyDir: {}
    name: credential-socket
  - emptyDir: {}
    name: workload-certs
  - emptyDir:
      medium: Memory
    name: istio-envoy
  - emptyDir: {}
    name: istio-data
  - downwardAPI:
      defaultMode: 420
      items:
      - fieldRef:
          apiVersion: v1
          fieldPath: metadata.labels
        path: labels
      - fieldRef:
          apiVersion: v1
          fieldPath: metadata.annotations
        path: annotations
    name: istio-podinfo
  - name: istio-token
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          audience: istio-ca
          expirationSeconds: 43200
          path: istio-token
  - configMap:
      defaultMode: 420
      name: istio-ca-root-cert
    name: istiod-ca-cert
  - name: kube-api-access-wknsc
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2023-04-12T05:27:25Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2023-04-12T05:27:26Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2023-04-12T05:27:26Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2023-04-12T05:27:23Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: containerd://9f2762d4f98efa5cde0070c6679fabbfcf099c450aba1a0b72f073eacd7c607b
    image: docker.io/istio/proxyv2:1.17.1
    imageID: docker.io/istio/proxyv2@sha256:2152aea5fbe2de20f08f3e0412ad7a4cd54a492240ff40974261ee4bdb43871d
    lastState: {}
    name: istio-proxy
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2023-04-12T05:27:25Z"
  - containerID: containerd://e7e2a70b7a4632a9105d58cd7f19d9f5d5a7c84621027172dddf8e2ba39e2461
    image: docker.io/library/nginx:latest
    imageID: docker.io/library/nginx@sha256:82c42833dbed48dc403246c532caef3aa7f5c6b5633d74a17ae565745f7215e9
    lastState: {}
    name: nginx
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2023-04-12T05:27:25Z"
  hostIP: 10.128.15.201
  initContainerStatuses:
  - containerID: containerd://f4452f732aa8fea3145b5281b55a097cb8c7f12c319162349baf31727a26e38c
    image: docker.io/istio/proxyv2:1.17.1
    imageID: docker.io/istio/proxyv2@sha256:2152aea5fbe2de20f08f3e0412ad7a4cd54a492240ff40974261ee4bdb43871d
    lastState: {}
    name: istio-init
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: containerd://f4452f732aa8fea3145b5281b55a097cb8c7f12c319162349baf31727a26e38c
        exitCode: 0
        finishedAt: "2023-04-12T05:27:24Z"
        reason: Completed
        startedAt: "2023-04-12T05:27:24Z"
  phase: Running
  podIP: 10.4.3.5
  podIPs:
  - ip: 10.4.3.5
  qosClass: Burstable
  startTime: "2023-04-12T05:27:23Z"

```
</details>

#### With the annotation
These are the local storage volumes present on the pod
![image](https://user-images.githubusercontent.com/34534103/231361429-42468137-0fe5-4992-8d70-8f7f5cdc6e1f.png)

Added annotation
![image](https://user-images.githubusercontent.com/34534103/231364409-fc19b6eb-5f18-46ee-802a-476d8df58267.png)
![image](https://user-images.githubusercontent.com/34534103/231364505-5b2cbf6a-7423-4c68-930b-e7b93d746da4.png)

Node is being scaled down
![image](https://user-images.githubusercontent.com/34534103/231365127-f72ae28f-51e7-4784-9060-d8a1ec029273.png)

<details>
<summary>
Here's what the pod looks like
</summary>

```
suraj@suraj:~/bin$ k get po nginx-7bd5c976c7-tctrh -oyaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes: workload-socket,credential-socket,workload-certs,istio-envoy,istio-data
    kubectl.kubernetes.io/default-container: nginx
    kubectl.kubernetes.io/default-logs-container: nginx
    prometheus.io/path: /stats/prometheus
    prometheus.io/port: "15020"
    prometheus.io/scrape: "true"
    sidecar.istio.io/status: '{"initContainers":["istio-init"],"containers":["istio-proxy"],"volumes":["workload-socket","credential-socket","workload-certs","istio-envoy","istio-data","istio-podinfo","istio-token","istiod-ca-cert"],"imagePullSecrets":null,"revision":"default"}'
  creationTimestamp: "2023-04-12T05:57:32Z"
  generateName: nginx-7bd5c976c7-
  labels:
    app: nginx
    pod-template-hash: 7bd5c976c7
    security.istio.io/tlsMode: istio
    service.istio.io/canonical-name: nginx
    service.istio.io/canonical-revision: latest
  name: nginx-7bd5c976c7-tctrh
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: nginx-7bd5c976c7
    uid: 53caba4f-d437-4e53-b108-ac2ac5a8a7d5
  resourceVersion: "53326"
  uid: 5607da02-69d1-452c-b049-3348f69696d0
spec:
  containers:
  - image: nginx
    imagePullPolicy: Always
    name: nginx
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-lfqxf
      readOnly: true
  - args:
    - proxy
    - sidecar
    - --domain
    - $(POD_NAMESPACE).svc.cluster.local
    - --proxyLogLevel=warning
    - --proxyComponentLogLevel=misc:error
    - --log_output_level=default:info
    - --concurrency
    - "2"
    env:
    - name: JWT_POLICY
      value: third-party-jwt
    - name: PILOT_CERT_PROVIDER
      value: istiod
    - name: CA_ADDR
      value: istiod.istio-system.svc:15012
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: INSTANCE_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    - name: SERVICE_ACCOUNT
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.serviceAccountName
    - name: HOST_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.hostIP
    - name: PROXY_CONFIG
      value: |
        {}
    - name: ISTIO_META_POD_PORTS
      value: |-
        [
        ]
    - name: ISTIO_META_APP_CONTAINERS
      value: nginx
    - name: ISTIO_META_CLUSTER_ID
      value: Kubernetes
    - name: ISTIO_META_NODE_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
    - name: ISTIO_META_INTERCEPTION_MODE
      value: REDIRECT
    - name: ISTIO_META_WORKLOAD_NAME
      value: nginx
    - name: ISTIO_META_OWNER
      value: kubernetes://apis/apps/v1/namespaces/default/deployments/nginx
    - name: ISTIO_META_MESH_ID
      value: cluster.local
    - name: TRUST_DOMAIN
      value: cluster.local
    image: docker.io/istio/proxyv2:1.17.1
    imagePullPolicy: IfNotPresent
    name: istio-proxy
    ports:
    - containerPort: 15090
      name: http-envoy-prom
      protocol: TCP
    readinessProbe:
      failureThreshold: 30
      httpGet:
        path: /healthz/ready
        port: 15021
        scheme: HTTP
      initialDelaySeconds: 1
      periodSeconds: 2
      successThreshold: 1
      timeoutSeconds: 3
    resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 10m
        memory: 40Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsGroup: 1337
      runAsNonRoot: true
      runAsUser: 1337
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/workload-spiffe-uds
      name: workload-socket
    - mountPath: /var/run/secrets/credential-uds
      name: credential-socket
    - mountPath: /var/run/secrets/workload-spiffe-credentials
      name: workload-certs
    - mountPath: /var/run/secrets/istio
      name: istiod-ca-cert
    - mountPath: /var/lib/istio/data
      name: istio-data
    - mountPath: /etc/istio/proxy
      name: istio-envoy
    - mountPath: /var/run/secrets/tokens
      name: istio-token
    - mountPath: /etc/istio/pod
      name: istio-podinfo
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-lfqxf
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  initContainers:
  - args:
    - istio-iptables
    - -p
    - "15001"
    - -z
    - "15006"
    - -u
    - "1337"
    - -m
    - REDIRECT
    - -i
    - '*'
    - -x
    - ""
    - -b
    - '*'
    - -d
    - 15090,15021,15020
    - --log_output_level=default:info
    image: docker.io/istio/proxyv2:1.17.1
    imagePullPolicy: IfNotPresent
    name: istio-init
    resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 10m
        memory: 40Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        add:
        - NET_ADMIN
        - NET_RAW
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: false
      runAsGroup: 0
      runAsNonRoot: false
      runAsUser: 0
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-lfqxf
      readOnly: true
  nodeName: gke-cluster-1-default-pool-87aace2f-vwr9
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir: {}
    name: workload-socket
  - emptyDir: {}
    name: credential-socket
  - emptyDir: {}
    name: workload-certs
  - emptyDir:
      medium: Memory
    name: istio-envoy
  - emptyDir: {}
    name: istio-data
  - downwardAPI:
      defaultMode: 420
      items:
      - fieldRef:
          apiVersion: v1
          fieldPath: metadata.labels
        path: labels
      - fieldRef:
          apiVersion: v1
          fieldPath: metadata.annotations
        path: annotations
    name: istio-podinfo
  - name: istio-token
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          audience: istio-ca
          expirationSeconds: 43200
          path: istio-token
  - configMap:
      defaultMode: 420
      name: istio-ca-root-cert
    name: istiod-ca-cert
  - name: kube-api-access-lfqxf
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2023-04-12T05:57:34Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2023-04-12T05:57:35Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2023-04-12T05:57:35Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2023-04-12T05:57:32Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: containerd://56f5be6512dcb1edf444e06cbcf1e09e7ca55d9c1bce7f126f0a795b77d69cec
    image: docker.io/istio/proxyv2:1.17.1
    imageID: docker.io/istio/proxyv2@sha256:2152aea5fbe2de20f08f3e0412ad7a4cd54a492240ff40974261ee4bdb43871d
    lastState: {}
    name: istio-proxy
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2023-04-12T05:57:34Z"
  - containerID: containerd://248f080691804d591428bcda1f4546e83eec20df1ab16c96eeb5a88c23b540e0
    image: docker.io/library/nginx:latest
    imageID: docker.io/library/nginx@sha256:82c42833dbed48dc403246c532caef3aa7f5c6b5633d74a17ae565745f7215e9
    lastState: {}
    name: nginx
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2023-04-12T05:57:34Z"
  hostIP: 10.128.15.201
  initContainerStatuses:
  - containerID: containerd://464130fec23a9688f1355436b2a93ca5dd8dac859a4c5ad1127395bccee2e67a
    image: docker.io/istio/proxyv2:1.17.1
    imageID: docker.io/istio/proxyv2@sha256:2152aea5fbe2de20f08f3e0412ad7a4cd54a492240ff40974261ee4bdb43871d
    lastState: {}
    name: istio-init
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: containerd://464130fec23a9688f1355436b2a93ca5dd8dac859a4c5ad1127395bccee2e67a
        exitCode: 0
        finishedAt: "2023-04-12T05:57:33Z"
        reason: Completed
        startedAt: "2023-04-12T05:57:33Z"
  phase: Running
  podIP: 10.4.3.6
  podIPs:
  - ip: 10.4.3.6
  qosClass: Burstable
  startTime: "2023-04-12T05:57:32Z"
```
</details>

#### CA params for both the cases above
![image](https://user-images.githubusercontent.com/34534103/231361120-8cb39634-8f3e-4115-906e-36ca90f9bac9.png)

vadasambar commented:
@towca I think I'm done from my end. Waiting for your review.

towca commented:
Everything looks good, thanks for the contribution!

I have one more small nit if you agree and want to fix, otherwise feel free to unhold the PR.

/lgtm
/approve
/hold

vadasambar commented:
> Everything looks good, thanks for the contribution!
> 
> I have one more small nit if you agree and want to fix, otherwise feel free to unhold the PR.
> 
> /lgtm /approve /hold

Thank you for the review. I have updated the PR to address the last comment.

vadasambar commented:
/unhold

vadasambar commented:
/lgtm

vadasambar commented:
@towca need a `/lgtm` to merge the PR. It was removed automatically after I updated the PR to address the last comment.

towca commented:
Thanks again!

/lgtm
/approve

jonathan-mothership commented:
Nice feature @vadasambar!

@MaciekPytel any idea on timeline for this to get released? It doesn't seem to be in latest which was cut about a month ago.

vadasambar commented:
> Nice feature @vadasambar!
> 
> @MaciekPytel any idea on timeline for this to get released? It doesn't seem to be in latest which was cut about a month ago. 

You can find it in 1.27 which will be released soon (check `cluster-autoscaler-release-1.27` branch. For other versions, we might have to backport.

thesuperzapper commented:
@vadasambar @towca what do we think about allowing `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` to be set as `"*"`, so that all local volumes for a Pod will be ignored, rather than having to specify them all?

vadasambar commented:
> @vadasambar @towca what do we think about allowing `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` to be set as `"*"`, so that all local volumes for a Pod will be ignored, rather than having to specify them all?

I think there is a room for improvement in the annotation.  `"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "*"` will have the same effect as `"cluster-autoscaler.kubernetes.io/safe-to-evict": "true"` ([ref](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node)). There are 2 cases here:
1. You have set `"cluster-autoscaler.kubernetes.io/safe-to-evict": "false"` (pod is not safe to evict)
2. You have set  `"cluster-autoscaler.kubernetes.io/safe-to-evict": "true"` (pod is safe to evict; if this is set `"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes"` annotation will be ignored)

In 1, if you add `"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "*"`, it will have the same effect as setting 2
In 2, if you add `"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "*"`, it will have the same effect as setting 2

Unless there is a specific usecase, I don't see a lot of value in adding  `"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "*"` because we already support `"cluster-autoscaler.kubernetes.io/safe-to-evict": "true"`. 

I guess what you want is  `"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "<container-name>:*"` e.g., `"cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "istio:*"` which makes sense because it provides better user experience than specifying the entire list.
